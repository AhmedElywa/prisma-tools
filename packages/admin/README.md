# @paljs/admin

A comprehensive React admin UI package for Prisma-based applications. This package provides ready-to-use components for building admin interfaces with full CRUD operations, form handling, and table management.

## Installation

```bash
npm install @paljs/admin
# or
yarn add @paljs/admin
# or
pnpm add @paljs/admin
```

## Dependencies

This package requires the following peer dependencies:
- React ^18.2.0
- React DOM ^18.2.0
- React Hook Form
- @apollo/client (for GraphQL operations)

## Features

- üé® **Modern UI Components** - Built with Tailwind CSS and Headless UI
- üìù **Dynamic Forms** - Auto-generated forms based on Prisma schema
- üìä **Data Tables** - Sortable, filterable tables with pagination
- üîß **CRUD Operations** - Complete Create, Read, Update, Delete functionality
- üéØ **Type Safety** - Full TypeScript support
- üé® **Customizable** - Flexible theming and styling options
- üì± **Responsive** - Mobile-friendly design

## Main Exports

### Components

#### PrismaTable
The main table component for displaying and managing Prisma model data.

```typescript
import { PrismaTable } from '@paljs/admin';

interface PrismaTableProps<T = any> {
  model: string;
  data: T[];
  columns: Column<T>[];
  onEdit?: (record: T) => void;
  onDelete?: (record: T) => void;
  onCreate?: () => void;
  loading?: boolean;
  pagination?: {
    page: number;
    pageSize: number;
    total: number;
    onPageChange: (page: number) => void;
    onPageSizeChange: (pageSize: number) => void;
  };
}

// Usage
<PrismaTable
  model="User"
  data={users}
  columns={userColumns}
  onEdit={handleEdit}
  onDelete={handleDelete}
  onCreate={handleCreate}
  loading={loading}
  pagination={paginationConfig}
/>
```

#### Settings
Configuration component for admin panel settings.

```typescript
import { Settings } from '@paljs/admin';

interface SettingsProps {
  models: ModelSettings[];
  onSave: (settings: ModelSettings[]) => void;
  onReset: () => void;
}

// Usage
<Settings
  models={modelSettings}
  onSave={handleSaveSettings}
  onReset={handleResetSettings}
/>
```

### Types and Interfaces

#### TableParentRecord
Interface for parent record relationships in tables.

```typescript
interface TableParentRecord {
  name: string;
  value: any;
  field: string;
  updateRecord?: () => Promise<any>;
}
```

#### Column Configuration
Define table columns with sorting, filtering, and custom rendering.

```typescript
interface Column<T> {
  key: keyof T;
  title: string;
  sortable?: boolean;
  filterable?: boolean;
  render?: (value: any, record: T) => React.ReactNode;
  width?: string | number;
}
```

### Form Components

#### Dynamic Form Generation
Auto-generate forms based on Prisma schema definitions.

```typescript
import { PrismaForm } from '@paljs/admin';

interface PrismaFormProps {
  model: string;
  fields: FieldConfig[];
  initialValues?: Record<string, any>;
  onSubmit: (values: Record<string, any>) => void;
  onCancel?: () => void;
  loading?: boolean;
}

// Usage
<PrismaForm
  model="User"
  fields={userFields}
  initialValues={editingUser}
  onSubmit={handleSubmit}
  onCancel={handleCancel}
  loading={submitting}
/>
```

### Utility Functions

#### classNames
Utility function for conditional CSS class names.

```typescript
function classNames(...classes: string[]): string;

// Usage
const buttonClass = classNames(
  'btn',
  'btn-primary',
  isLoading && 'btn-loading',
  isDisabled && 'btn-disabled'
);
```

## Advanced Usage

### Custom Table Columns

```typescript
const userColumns: Column<User>[] = [
  {
    key: 'id',
    title: 'ID',
    sortable: true,
    width: 80,
  },
  {
    key: 'name',
    title: 'Name',
    sortable: true,
    filterable: true,
  },
  {
    key: 'email',
    title: 'Email',
    sortable: true,
    filterable: true,
  },
  {
    key: 'createdAt',
    title: 'Created',
    sortable: true,
    render: (value) => new Date(value).toLocaleDateString(),
  },
  {
    key: 'actions',
    title: 'Actions',
    render: (_, record) => (
      <div className="flex space-x-2">
        <button onClick={() => handleEdit(record)}>Edit</button>
        <button onClick={() => handleDelete(record)}>Delete</button>
      </div>
    ),
  },
];
```

### Form Field Configuration

```typescript
const userFields: FieldConfig[] = [
  {
    name: 'name',
    type: 'text',
    label: 'Full Name',
    required: true,
    validation: {
      minLength: 2,
      maxLength: 50,
    },
  },
  {
    name: 'email',
    type: 'email',
    label: 'Email Address',
    required: true,
    validation: {
      pattern: /^[^\s@]+@[^\s@]+\.[^\s@]+$/,
    },
  },
  {
    name: 'role',
    type: 'select',
    label: 'Role',
    options: [
      { value: 'USER', label: 'User' },
      { value: 'ADMIN', label: 'Admin' },
    ],
  },
];
```

### Integration with Apollo Client

```typescript
import { useQuery, useMutation } from '@apollo/client';
import { PrismaTable } from '@paljs/admin';

function UserManagement() {
  const { data, loading, refetch } = useQuery(GET_USERS);
  const [deleteUser] = useMutation(DELETE_USER);
  const [updateUser] = useMutation(UPDATE_USER);

  const handleDelete = async (user: User) => {
    await deleteUser({ variables: { id: user.id } });
    refetch();
  };

  const handleEdit = async (user: User) => {
    // Open edit modal or navigate to edit page
  };

  return (
    <PrismaTable
      model="User"
      data={data?.users || []}
      columns={userColumns}
      onEdit={handleEdit}
      onDelete={handleDelete}
      loading={loading}
    />
  );
}
```

## Styling

The package uses Tailwind CSS for styling. Make sure to include Tailwind CSS in your project:

```bash
npm install -D tailwindcss @tailwindcss/forms
```

Include the admin styles in your CSS:

```css
@import '@paljs/admin/dist/style.css';
```

## TypeScript Support

This package is written in TypeScript and provides full type definitions. All components and functions are properly typed for the best development experience.

## Browser Support

- Chrome (latest)
- Firefox (latest)
- Safari (latest)
- Edge (latest)

## Contributing

This package is part of the PalJS ecosystem. For contributing guidelines, please refer to the main repository.

## License

MIT License - see the LICENSE file for details.
