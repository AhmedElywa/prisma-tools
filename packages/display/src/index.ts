import chalk from 'chalk';
import ora from 'ora';
import readline from 'readline';
import debug from 'debug';

const debugName = 'paljs';

// Using brigh brand color, so it's better for dark terminals
const brandColor = '7ca2ff';

const withBrand = (str: string) => {
  return chalk.hex(brandColor).bold(str);
};

const withWarning = (str: string) => {
  return `⚠️  ${chalk.yellow(str)}`;
};

const withCaret = (str: string) => {
  return `${chalk.gray('>')} ${str}`;
};

const withCheck = (str: string) => {
  return `${chalk.green('✔')} ${str}`;
};

const withX = (str: string) => {
  return `${chalk.red.bold('✕')} ${str}`;
};

/**
 * Logs a branded purple message to stdout.
 *
 * @param {string} msg
 */
const branded = (msg: string) => {
  console.log(chalk.hex(brandColor).bold(msg));
};

/**
 * Clears the line and optionally log a message to stdout.
 *
 * @param {string} msg
 */
const clearLine = (msg?: string) => {
  readline.clearLine(process.stdout, 0);
  readline.cursorTo(process.stdout, 0);
  void (msg && process.stdout.write(msg));
};

/**
 * Logs a red error message to stderr.
 *
 * @param {string} msg
 */
const warning = (msg: string) => {
  console.log(withCaret(withWarning(msg)));
};

/**
 * Logs a red error message to stderr.
 *
 * @param {string} msg
 */
const error = (msg: string) => {
  console.error(withX(chalk.red.bold(msg)));
};

/**
 * Logs a subtle gray message to stdout.
 *
 * @param {string} msg
 */
const meta = (msg: string) => {
  console.log(withCaret(chalk.gray(msg)));
};

/**
 * Logs a progress message to stdout.
 *
 * @param {string} msg
 */
const progress = (msg: string) => {
  console.log(withCaret(chalk.bold(msg)));
};

const info = (msg: string) => {
  console.log(chalk.bold(msg));
};

const spinner = (str: string) => {
  return ora({
    text: str,
    color: 'blue',
    spinner: {
      interval: 120,
      frames: ['◢', '◣', '◤', '◥'],
    },
  });
};

/**
 * Logs a green success message to stdout.
 *
 * @param {string} msg
 */
const success = (msg: string) => {
  console.log(withCheck(chalk.green(msg)));
};

const newline = () => {
  console.log(' ');
};

/**
 * Colorizes a variable for display.
 *
 * @param {string} val
 */
const variable = (val: string) => {
  return chalk.cyan.bold(`${val}`);
};

const d = debug(debugName);

const throwError = (str: string) => {
  throw new Error(removeColorInTest(withX(str)));
};

const removeColorInTest = (str: string) => {
  // eslint-disable-next-line no-control-regex
  return process.env.NODE_ENV === 'test' ? str.replace(/\x1b\[[0-9]+m/g, '') : str;
};

export const log = {
  withBrand,
  withWarning,
  withCaret,
  withCheck,
  withX,
  branded,
  clearLine,
  error,
  warning,
  meta,
  progress,
  spinner,
  success,
  newline,
  variable,
  info,
  d,
  removeColorInTest,
  throwError,
};

export { chalk };
