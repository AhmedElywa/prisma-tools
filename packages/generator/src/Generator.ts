import { GeneratorsType, GeneratorOptions } from '@paljs/types';
import { GenerateNexus } from './nexus';
import { GenerateSdl } from './sdl';
import { GenerateModules } from './graphql-modules';

export class Generator {
  generators: Record<GeneratorsType, GenerateNexus | GenerateSdl | GenerateModules>;

  constructor(
    private generator: { name: GeneratorsType; schemaPath: string },
    private options?: Partial<GeneratorOptions>,
  ) {
    this.generators = {
      nexus: new GenerateNexus(this.generator.schemaPath, this.options),
      sdl: new GenerateSdl(this.generator.schemaPath, this.options),
      'graphql-modules': new GenerateModules(this.generator.schemaPath, this.options),
    };
  }

  async run() {
    if (this.generators[this.generator.name]) {
      await this.generators[this.generator.name].run();
    } else {
      console.error(
        `Your generator name: "${this.generator.name}" not correct.\nPlease use one of this`,
        Object.keys(this.generators),
      );
    }
  }
}
