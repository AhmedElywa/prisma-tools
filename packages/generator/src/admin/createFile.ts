import { writeFileSync, mkdirSync, existsSync } from 'fs';
import { join } from 'path';

export const createFile = (path: string, fileName: string, content: string) => {
  void (!existsSync(path) && mkdirSync(path, { recursive: true }));
  void (!existsSync(join(path, fileName)) && writeFileSync(join(path, fileName), content));
};
