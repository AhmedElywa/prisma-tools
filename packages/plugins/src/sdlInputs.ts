import { GraphQLSchema, printSchema } from 'graphql';
import { writeFileSync } from 'fs';

export const generateGraphQlSDLFile = (schema: GraphQLSchema, path = 'schema.graphql') => {
  writeFileSync(path, printSchema(schema));
};
