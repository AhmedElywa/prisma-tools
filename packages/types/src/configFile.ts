import { GeneratorOptions } from './generator';

export type RouterType = 'pages' | 'app';

export interface AdminPagesOptions<ModelName extends string = string> {
  models?: ModelName[];
  pageContent?: string;
  outPut?: string;
  backAsText?: boolean;
  routerType?: RouterType;
}

export type PartialOptions<
  ModelName extends string = string,
  ModelsObject extends Record<ModelName, Record<string, any>> = Record<ModelName, Record<string, any>>,
> = Omit<Partial<GeneratorOptions<ModelName, ModelsObject>>, 'backAsText'>;

export type GeneratorsType = 'nexus' | 'sdl' | 'graphql-modules';

export interface Config<
  ModelName extends string = string,
  ModelsObject extends Record<ModelName, Record<string, any>> = Record<ModelName, Record<string, any>>,
> {
  schema?: string;
  backend?: {
    generator: GeneratorsType;
    adminSettingsPath?: string;
  } & PartialOptions<ModelName, ModelsObject>;
  frontend?: {
    admin?: AdminPagesOptions<ModelName> | boolean;
    graphql?: PartialOptions<ModelName, ModelsObject> | boolean;
  };
}

export type MultiSchemaConfig<
  ModelName extends string = string,
  ModelsObject extends Record<ModelName, Record<string, any>> = Record<ModelName, Record<string, any>>,
> = Record<string, Config<ModelName, ModelsObject>>;
