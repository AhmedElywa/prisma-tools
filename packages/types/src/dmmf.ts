export type { DMMF } from '@prisma/generator-helper';

export type ReadonlyDeep<O> = {
  +readonly [K in keyof O]: ReadonlyDeep<O[K]>;
};
