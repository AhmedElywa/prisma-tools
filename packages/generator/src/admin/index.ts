import { writeFileSync } from 'fs';
import { format } from 'prettier';
import { GeneratorOptions, AdminPagesOptions, SchemaObject } from '@paljs/types';
import { createGraphql } from './createGraphql';
import { mergeSchema } from './mergeSchema';
import { ConvertSchemaToObject } from '@paljs/schema';
import { createFile } from './createFile';

const defaultOptions: GeneratorOptions = {
  prismaName: 'prisma',
  output: './src/graphql',
  excludeFields: [],
  excludeModels: [],
  excludeFieldsByModel: {},
  excludeQueriesAndMutations: [],
  excludeQueriesAndMutationsByModel: {},
};

export class UIGenerator {
  schema: SchemaObject;

  constructor(schemaPath: string | string[]) {
    this.schema =
      typeof schemaPath === 'string'
        ? new ConvertSchemaToObject(schemaPath).run()
        : this.mergeSchemas(schemaPath.map((path) => new ConvertSchemaToObject(path).run()));
  }

  mergeSchemas(schemas: SchemaObject[]) {
    const schema: SchemaObject = { models: [], enums: [] };
    schemas.forEach((s) => {
      schema.models.push(...s.models);
      schema.enums.push(...s.enums);
    });
    return schema;
  }

  async buildSettingsSchema(path = 'adminSettings.json', backAsText?: boolean) {
    const newSchema = mergeSchema(this.schema, path);
    const fileContent = await format(`${JSON.stringify(newSchema)}`, {
      singleQuote: true,
      semi: false,
      trailingComma: 'all',
      parser: 'json',
    });

    void (!backAsText && writeFileSync(path, fileContent));

    return newSchema;
  }

  async generateGraphql(customOptions?: Omit<Partial<GeneratorOptions>, 'nexusSchema'>) {
    const options: GeneratorOptions = {
      ...defaultOptions,
      ...customOptions,
    };
    return await createGraphql(this.schema, options);
  }

  /**
   * Generate admin pages for Prisma models
   * @param options - Configuration options for page generation
   * @param options.routerType - Router type: 'pages' for Next.js Pages Router or 'app' for Next.js App Router
   * @param options.models - Array of model names to generate pages for
   * @param options.pageContent - Custom page template content
   * @param options.outPut - Output directory path
   * @param options.backAsText - Return generated content as text instead of writing files
   * @returns Generated page content as text (if backAsText is true) or void
   */
  async generateAdminPages(options?: AdminPagesOptions) {
    const routerType = options?.routerType ?? 'pages';
    const content = options?.pageContent ?? (routerType === 'app' ? appRouterPage : pagesRouterPage);
    const generatedText: Record<string, string> = {};
    const filterModels = this.schema.models.filter((model) => !options?.models || options?.models.includes(model.name));

    // Generate layout files for App Router if needed
    if (routerType === 'app' && !options?.backAsText) {
      await this.generateAppRouterLayouts(options);
    }

    for (const model of filterModels) {
      if (routerType === 'app') {
        // Generate App Router structure: app/admin/models/[model]/page.tsx
        const fileContent = await format(content.replace(/#{id}/g, model.name), {
          semi: true,
          trailingComma: 'all',
          singleQuote: true,
          printWidth: 120,
          tabWidth: 2,
          parser: 'babel-ts',
        });

        if (options?.backAsText) {
          generatedText[model.name] = fileContent;
        } else {
          const outputPath = options?.outPut ?? 'src/app/admin/models/';
          const modelPath = `${outputPath}${model.name}/`;
          createFile(modelPath, 'page.tsx', fileContent);
        }
      } else {
        // Generate Pages Router structure: pages/admin/models/[model].tsx
        const fileContent = await format(content.replace(/#{id}/g, model.name), {
          semi: true,
          trailingComma: 'all',
          singleQuote: true,
          printWidth: 120,
          tabWidth: 2,
          parser: 'babel-ts',
        });

        if (options?.backAsText) {
          generatedText[model.name] = fileContent;
        } else {
          createFile(options?.outPut ?? 'src/pages/admin/models/', `${model.name}.tsx`, fileContent);
        }
      }
    }
    return generatedText;
  }

  async generateAppRouterLayouts(options?: AdminPagesOptions) {
    const outputPath = options?.outPut ?? 'src/app/admin/';

    // Generate root admin layout
    const adminLayoutContent = await format(adminLayout, {
      semi: true,
      trailingComma: 'all',
      singleQuote: true,
      printWidth: 120,
      tabWidth: 2,
      parser: 'babel-ts',
    });
    createFile(outputPath, 'layout.tsx', adminLayoutContent);

    // Generate models layout
    const modelsLayoutContent = await format(modelsLayout, {
      semi: true,
      trailingComma: 'all',
      singleQuote: true,
      printWidth: 120,
      tabWidth: 2,
      parser: 'babel-ts',
    });
    createFile(`${outputPath}models/`, 'layout.tsx', modelsLayoutContent);
  }
}

// Pages Router template (existing)
const pagesRouterPage = `
import React from 'react';
import PrismaTable from 'components/PrismaTable';

const #{id}: React.FC = () => {
  return <PrismaTable model="#{id}" />;
};

export default #{id};
`;

// App Router template (new)
const appRouterPage = `
import React from 'react';
import PrismaTable from 'components/PrismaTable';

export default function #{id}Page() {
  return <PrismaTable model="#{id}" />;
}
`;

// App Router admin layout template
const adminLayout = `
import React from 'react';

export default function AdminLayout({
  children,
}: {
  children: React.ReactNode;
}) {
  return (
    <div className="admin-layout">
      {children}
    </div>
  );
}
`;

// App Router models layout template
const modelsLayout = `
import React from 'react';

export default function ModelsLayout({
  children,
}: {
  children: React.ReactNode;
}) {
  return (
    <div className="models-layout">
      {children}
    </div>
  );
}
`;
