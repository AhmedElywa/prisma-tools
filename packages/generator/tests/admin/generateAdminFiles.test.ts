import { join } from 'path';
import { UIGenerator } from '../../src/admin';

describe('Generate Admin Files', () => {
  const schemaPath = join(__dirname, '../schemas/schema.prisma');
  it('Should back with models pages', async () => {
    const generator = new UIGenerator(schemaPath);
    const text = await generator.generateAdminPages({ backAsText: true });
    expect(text).toMatchSnapshot();
  });

  test('Generate admin files for App Router', async () => {
    const uiGenerator = new UIGenerator('./packages/generator/tests/schemas/schema.prisma');
    const result = await uiGenerator.generateAdminPages({
      models: ['User'],
      backAsText: true,
      routerType: 'app',
    });
    expect(result).toMatchSnapshot();
  });

  test('Generate admin files for Pages Router explicitly', async () => {
    const uiGenerator = new UIGenerator('./packages/generator/tests/schemas/schema.prisma');
    const result = await uiGenerator.generateAdminPages({
      models: ['User'],
      backAsText: true,
      routerType: 'pages',
    });
    expect(result).toMatchSnapshot();
  });
});
