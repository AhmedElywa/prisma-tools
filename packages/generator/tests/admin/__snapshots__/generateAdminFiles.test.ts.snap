// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generate Admin Files Generate admin files for App Router 1`] = `
{
  "User": "import React from 'react';
import PrismaTable from 'components/PrismaTable';

export default function UserPage() {
  return <PrismaTable model="User" />;
}
",
}
`;

exports[`Generate Admin Files Generate admin files for Pages Router explicitly 1`] = `
{
  "User": "import React from 'react';
import PrismaTable from 'components/PrismaTable';

const User: React.FC = () => {
  return <PrismaTable model="User" />;
};

export default User;
",
}
`;

exports[`Generate Admin Files Should back with models pages 1`] = `
{
  "Post": "import React from 'react';
import PrismaTable from 'components/PrismaTable';

const Post: React.FC = () => {
  return <PrismaTable model="Post" />;
};

export default Post;
",
  "User": "import React from 'react';
import PrismaTable from 'components/PrismaTable';

const User: React.FC = () => {
  return <PrismaTable model="User" />;
};

export default User;
",
}
`;
